pkg/bgp/speaker/events.go

```diff
@@ -49,18 +69,57 @@ func (s *Speaker) run() {
 // do performs the appropriate action depending on the event type. For example,
 // if it is a service event (svcEvent), then it will call into MetalLB's
 // SetService() to perform BGP announcements.
-func (s *Speaker) do(key interface{}) types.SyncState {
+func (s *MetalLBSpeaker) do(key interface{}) types.SyncState {
 	defer s.queue.Done(key)
 
 	switch k := key.(type) {
 	case svcEvent:
-		return s.SetService(s.logger, k.id.String(), k.svc, k.eps)
+		return s.speaker.SetService(k.id.String(), k.svc, k.eps)
 	case epEvent:
-		return s.SetService(s.logger, k.id.String(), k.svc, k.eps)
+		return s.speaker.SetService(k.id.String(), k.svc, k.eps)
 	case nodeEvent:
-		return s.SetNodeLabels(s.logger, *k)
+		return s.handleNodeEvent(k)
 	default:
 		log.Debugf("Encountered an unknown key type %T in BGP speaker", k)
 		return types.SyncStateSuccess
 	}
 }
+
+func (s *MetalLBSpeaker) handleNodeEvent(k nodeEvent) types.SyncState {
+	var (
+		ret    types.SyncState
+		failed bool
+	)
+
+	if k.withDraw {
+		// this is a best effort method call, so we don't
+		// care about errors. If the node is shutting down
+		// all routes will be closed once the speaker's TCP conn
+		// is closed anyway per the rfc.
+		// see: https://datatracker.ietf.org/doc/html/rfc4271#section-6
+		s.withDraw()
+		atomic.AddInt32(&s.shutdown, 1)
+		return types.SyncStateSuccess
+	}
+
+	if s.announceLBIP {
+		if r := s.speaker.SetNodeLabels(*k.labels); r != types.SyncStateSuccess {
+			failed = true
```

Comment:
Should this be logged as in the case below?

Reply:
I would think so, fixing this :+1: